kv/kvserver: TestMergeQueue failed

kv/kvserver.TestMergeQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3788880&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3788880&tab=artifacts#/) on master @ [a132b15e6c9705a6922c2e476936597c6670e072](https://github.com/cockroachdb/cockroach/commits/a132b15e6c9705a6922c2e476936597c6670e072):


```
=== RUN   TestMergeQueue
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue870837684
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestMergeQueue
    client_merge_test.go:4453: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue870837684
--- FAIL: TestMergeQueue (2.58s)
=== RUN   TestMergeQueue/load-based-merging/combined-qps-below-threshold
    client_merge_test.go:4397: ranges unexpectedly unmerged expected startKey /Table/Max, but got "\xfa\x00\x00"
        --- FAIL: TestMergeQueue/load-based-merging/combined-qps-below-threshold (0.03s)
=== RUN   TestMergeQueue/load-based-merging
    --- FAIL: TestMergeQueue/load-based-merging (0.14s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-11429

kv/kvserver.TestMergeQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3875241&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3875241&tab=artifacts#/) on master @ [a96f5b255fead76a94759231a0ee9b699e810956](https://github.com/cockroachdb/cockroach/commits/a96f5b255fead76a94759231a0ee9b699e810956):


```
=== RUN   TestMergeQueue
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue3321291283
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestMergeQueue
    client_merge_test.go:4453: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue3321291283
--- FAIL: TestMergeQueue (2.57s)
=== RUN   TestMergeQueue/load-based-merging/combined-qps-below-threshold
    client_merge_test.go:4397: ranges unexpectedly unmerged expected startKey /Table/Max, but got "\xfa\x00\x00"
        --- FAIL: TestMergeQueue/load-based-merging/combined-qps-below-threshold (0.03s)
=== RUN   TestMergeQueue/load-based-merging
    --- FAIL: TestMergeQueue/load-based-merging (0.14s)
```
<details><summary>Help</summary>
<p>
See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4141074&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4141074&tab=artifacts#/) on master @ [365b4da8bd02c06ee59d2130a56dec74ffc9ce21](https://github.com/cockroachdb/cockroach/commits/365b4da8bd02c06ee59d2130a56dec74ffc9ce21):


```
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:500 +0x109
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:481 +0x10a
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:311 +0x4b5
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:180 +0xa8
  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:4169 +0x6f8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47

Goroutine 659 (running) created at:
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x615
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:354 +0x13d
  github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigkvsubscriber.(*KVSubscriber).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigkvsubscriber/kvsubscriber.go:177 +0x26
  github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1801 +0x41fa
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:327 +0x161
  github.com/cockroachdb/cockroach/pkg/server.(*Node).writeNodeStatus()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:834 +0x104
  github.com/cockroachdb/cockroach/pkg/server.(*Node).startWriteNodeStatus()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:794 +0xae
  github.com/cockroachdb/cockroach/pkg/server.(*Server).PreStart()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1796 +0x416f
  github.com/cockroachdb/cockroach/pkg/server.(*Server).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:1251 +0x44
  github.com/cockroachdb/cockroach/pkg/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/server/testserver.go:500 +0x109
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).startServer()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:481 +0x10a
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.(*TestCluster).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:311 +0x4b5
  github.com/cockroachdb/cockroach/pkg/testutils/testcluster.StartTestCluster()
      /go/src/github.com/cockroachdb/cockroach/pkg/testutils/testcluster/testcluster.go:180 +0xa8
  github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestMergeQueue()
      /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/client_merge_test.go:4169 +0x6f8
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1259 +0x22f
  testing.(*T).Run·dwrap·21()
      /usr/local/go/src/testing/testing.go:1306 +0x47
==================
    testing.go:1152: race detected during execution of test
    --- FAIL: TestMergeQueue/sanity (1.41s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-race -parallel=2
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestMergeQueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4336521&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4336521&tab=artifacts#/) on master @ [456ff4c8692512f6d6f2b196f98c4a0691ca4589](https://github.com/cockroachdb/cockroach/commits/456ff4c8692512f6d6f2b196f98c4a0691ca4589):


```
=== RUN   TestMergeQueue
    test_log_scope.go:79: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue390597492
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestMergeQueue
    client_merge_test.go:4456: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestMergeQueue390597492
--- FAIL: TestMergeQueue (2.15s)
=== RUN   TestMergeQueue/load-based-merging/combined-qps-below-threshold
    client_merge_test.go:4400: ranges unexpectedly unmerged expected startKey /Table/Max, but got "\xfa\x00\x00"
        --- FAIL: TestMergeQueue/load-based-merging/combined-qps-below-threshold (0.02s)
=== RUN   TestMergeQueue/load-based-merging
    --- FAIL: TestMergeQueue/load-based-merging (0.04s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMergeQueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/client_merge_test.go
+++ b/pkg/kv/kvserver/client_merge_test.go
@@ -4143,26 +4143,36 @@ func TestStoreRangeMergeDuringShutdown(t *testing.T) {
 	}
 }
 
-func verifyMerged(t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey) {
+func verifyMergedSoon(t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey) {
 	t.Helper()
-	repl := store.LookupReplica(rhsStartKey)
-	if repl == nil {
-		t.Fatal("replica doesn't exist")
-	}
-	if !repl.Desc().StartKey.Equal(lhsStartKey) {
-		t.Fatalf("ranges unexpectedly unmerged expected startKey %s, but got %s", lhsStartKey, repl.Desc().StartKey)
-	}
+	testutils.SucceedsSoon(t, func() error {
+		store.MustForceMergeScanAndProcess()
+		repl := store.LookupReplica(rhsStartKey)
+		if repl == nil {
+			t.Fatal("replica doesn't exist")
+		}
+		if !repl.Desc().StartKey.Equal(lhsStartKey) {
+			return errors.Newf("ranges unexpectedly unmerged expected startKey %s, but got %s", lhsStartKey, repl.Desc().StartKey)
+		}
+		return nil
+	})
 }
 
-func verifyUnmerged(t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey) {
+func verifyUnmergedSoon(
+	t *testing.T, store *kvserver.Store, lhsStartKey, rhsStartKey roachpb.RKey,
+) {
 	t.Helper()
-	repl := store.LookupReplica(rhsStartKey)
-	if repl == nil {
-		t.Fatal("replica doesn't exist")
-	}
-	if repl.Desc().StartKey.Equal(lhsStartKey) {
-		t.Fatalf("ranges unexpectedly merged")
-	}
+	testutils.SucceedsSoon(t, func() error {
+		store.MustForceMergeScanAndProcess()
+		repl := store.LookupReplica(rhsStartKey)
+		if repl == nil {
+			t.Fatal("replica doesn't exist")
+		}
+		if repl.Desc().StartKey.Equal(lhsStartKey) {
+			return errors.Newf("ranges unexpectedly merged")
+		}
+		return nil
+	})
 }
 
 func TestMergeQueue(t *testing.T) {
@@ -4268,18 +4278,15 @@ func TestMergeQueue(t *testing.T) {
 	t.Run("sanity", func(t *testing.T) {
 		// Check that ranges are not trivially merged after reset.
 		reset(t)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 		reset(t)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("both-empty", func(t *testing.T) {
 		reset(t)
 		clearRange(t, lhsStartKey, rhsEndKey)
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("lhs-undersize", func(t *testing.T) {
@@ -4287,8 +4294,7 @@ func TestMergeQueue(t *testing.T) {
 		conf := conf
 		conf.RangeMinBytes *= 2
 		lhs().SetSpanConfig(conf)
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("combined-size-threshold", func(t *testing.T) {
@@ -4300,22 +4306,17 @@ func TestMergeQueue(t *testing.T) {
 		conf.RangeMinBytes = rhs().GetMVCCStats().Total() + 1
 		conf.RangeMaxBytes = lhs().GetMVCCStats().Total() + rhs().GetMVCCStats().Total() - 1
 		setSpanConfigs(t, conf)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Once the maximum size threshold is increased, the merge can occur.
 		conf.RangeMaxBytes += 1
 		setSpanConfigs(t, conf)
-		l := lhs().RangeID
-		r := rhs().RangeID
-		log.Infof(ctx, "Left=%s, Right=%s", l, r)
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("non-collocated", func(t *testing.T) {
 		reset(t)
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 		tc.AddVotersOrFatal(t, rhs().Desc().StartKey.AsRawKey(), tc.Target(1))
 		tc.TransferRangeLeaseOrFatal(t, *rhs().Desc(), tc.Target(1))
 		// NB: We're running on a 2 node `TestCluster` so we can count on the
@@ -4324,8 +4325,7 @@ func TestMergeQueue(t *testing.T) {
 		tc.RemoveVotersOrFatal(t, rhs().Desc().StartKey.AsRawKey(), tc.Target(0))
 
 		clearRange(t, lhsStartKey, rhsEndKey)
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("load-based-merging", func(t *testing.T) {
@@ -4371,8 +4371,7 @@ func TestMergeQueue(t *testing.T) {
 			lhs().LoadBasedSplitter().Reset(tc.Servers[0].Clock().PhysicalTime())
 
 			clearRange(t, lhsStartKey, rhsEndKey)
-			store.MustForceMergeScanAndProcess()
-			verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+			verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 		})
 
 		t.Run("unreliable-rhs-qps", func(t *testing.T) {
@@ -4381,8 +4380,7 @@ func TestMergeQueue(t *testing.T) {
 			rhs().LoadBasedSplitter().Reset(tc.Servers[1].Clock().PhysicalTime())
 
 			clearRange(t, lhsStartKey, rhsEndKey)
-			store.MustForceMergeScanAndProcess()
-			verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+			verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 		})
 
 		t.Run("combined-qps-above-threshold", func(t *testing.T) {
@@ -4393,8 +4391,7 @@ func TestMergeQueue(t *testing.T) {
 			lhs().LoadBasedSplitter().RecordMax(tc.Servers[1].Clock().PhysicalTime(), float64(moreThanHalfQPS))
 
 			clearRange(t, lhsStartKey, rhsEndKey)
-			store.MustForceMergeScanAndProcess()
-			verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+			verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 		})
 
 		t.Run("combined-qps-below-threshold", func(t *testing.T) {
@@ -4406,21 +4403,18 @@ func TestMergeQueue(t *testing.T) {
 			lhs().LoadBasedSplitter().RecordMax(tc.Servers[1].Clock().PhysicalTime(), float64(lessThanHalfQPS))
 
 			clearRange(t, lhsStartKey, rhsEndKey)
-			store.MustForceMergeScanAndProcess()
-			verifyMerged(t, store, lhsStartKey, rhsStartKey)
+			verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 		})
 	})
 
 	t.Run("sticky-bit", func(t *testing.T) {
 		reset(t)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Perform manual merge and verify that no merge occurred.
 		split(t, rhsStartKey.AsRawKey(), hlc.MaxTimestamp /* expirationTime */)
 		clearRange(t, lhsStartKey, rhsEndKey)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Delete sticky bit and verify that merge occurs.
 		unsplitArgs := &roachpb.AdminUnsplitRequest{
@@ -4431,15 +4425,13 @@ func TestMergeQueue(t *testing.T) {
 		if _, err := kv.SendWrapped(ctx, store.DB().NonTransactionalSender(), unsplitArgs); err != nil {
 			t.Fatal(err)
 		}
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 
 	t.Run("sticky-bit-expiration", func(t *testing.T) {
 		manualSplitTTL := 10 * time.Millisecond
 		reset(t)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Pause the clock for the duration of this subtest so that
 		// the sticky bit does not expire prematurely.
@@ -4450,18 +4442,15 @@ func TestMergeQueue(t *testing.T) {
 		exp := tc.Servers[0].Clock().Now().Add(manualSplitTTL.Nanoseconds(), 0)
 		split(t, rhsStartKey.AsRawKey(), exp /* expirationTime */)
 		clearRange(t, lhsStartKey, rhsEndKey)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Sticky bit is not expired yet.
 		manualClock.Increment(manualSplitTTL.Nanoseconds() / 2)
-		store.MustForceMergeScanAndProcess()
-		verifyUnmerged(t, store, lhsStartKey, rhsStartKey)
+		verifyUnmergedSoon(t, store, lhsStartKey, rhsStartKey)
 
 		// Sticky bit is expired.
 		manualClock.Increment(manualSplitTTL.Nanoseconds())
-		store.MustForceMergeScanAndProcess()
-		verifyMerged(t, store, lhsStartKey, rhsStartKey)
+		verifyMergedSoon(t, store, lhsStartKey, rhsStartKey)
 	})
 }
 
@@ -4663,8 +4652,7 @@ func TestMergeQueueSeesNonVoters(t *testing.T) {
 			rightDesc = tc.LookupRangeOrFatal(t, rightDesc.StartKey.AsRawKey())
 
 			store.SetMergeQueueActive(true)
-			store.MustForceMergeScanAndProcess()
-			verifyMerged(t, store, leftDesc.StartKey, rightDesc.StartKey)
+			verifyMergedSoon(t, store, leftDesc.StartKey, rightDesc.StartKey)
 		})
 	}
 }
@@ -4749,8 +4737,7 @@ func TestMergeQueueWithSlowNonVoterSnaps(t *testing.T) {
 		return nil
 	})
 	store.SetMergeQueueActive(true)
-	store.MustForceMergeScanAndProcess()
-	verifyMerged(t, store, leftDesc.StartKey, rightDesc.StartKey)
+	verifyMergedSoon(t, store, leftDesc.StartKey, rightDesc.StartKey)
 }
 
 func TestInvalidSubsumeRequest(t *testing.T) {

